mgmt: mcumgr: make smp_shell work with shell logs

ExtremeGTX opened the pull request:
This fix is for disabling shell_log_backend temporarily when MCUMGR
data exchange is ongoing. and reactivate when it finish.

This should fix #36882

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

ExtremeGTX commented:
@sw Could you please test this patch ?

ExtremeGTX commented on subsys/shell/shell_log_backend.c:
```diff
@@ -136,6 +136,18 @@ void z_shell_log_backend_disable(const struct shell_log_backend *backend)
+{
+	if (backend->control_block->state != SHELL_LOG_BACKEND_ENABLED) {
+		uint32_t level =
+			(CONFIG_SHELL_BACKEND_SERIAL_LOG_LEVEL > LOG_LEVEL_DBG) ?
```
Is there any more generic replacement for `CONFIG_SHELL_BACKEND_SERIAL_LOG_LEVEL` ?

ExtremeGTX replied:
@de-nordic

de-nordic replied:
The best solution, I can see for this, would be to save previous log level.
This means that we should probably modify `log_backend_disable` to return previous level, or define `log_backend_suspend` that would do that, and then use the level with `z_shell_log_backend_resume`.

ExtremeGTX replied:
But this also will store only one log level, instead of "per log source level", right ?

ExtremeGTX replied:
Do you know why `log_backend_disable` deactivate the backend and also change the loglevel(s) ?
I would expect that `log_backend_suspend` implementation will be to just deactivate the backend without changing the loglevel(s)

de-nordic replied:
Unfortunately I currently posses  "here be dragons" level of knowledge on internals of the log subsystem.

ExtremeGTX replied:
I found a way simpler solution to this problem, I will update this PR soon.

ExtremeGTX replied:
I updated the PR.
The solution is to just change the backend state `backend->control_block->state` between enabled/disabled.
This also will maintain the log level changes in the runtime.

sw replied:
I haven't looked into this closely, but it seems you're suspending logging before transmitting an SMP packet, but you only resume it on reception of a non-SMP byte.

Shouldn't there be a resume when the packet has been transmitted completely? Otherwise logging will stay disabled until we receive something.

de-nordic replied:
@ExtremeGTX  the mcumgr does have callback mechanism in img_mgmt; callbacks are registered with: `img_mgmt_register_callbacks`

de-nordic replied:
@ExtremeGTX Then propose new callbacks.

ExtremeGTX committed:
mgmt: mcumgr: make smp_shell work with shell logs

This fix is for disabling shell_log_backend temporarily when MCUMGR
data exchange is ongoing, and reactivate shell_log_backend again
when it finish.

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

lairdjm commented on subsys/mgmt/mcumgr/smp_shell.c:
```diff
@@ -85,6 +85,13 @@ static int read_mcumgr_byte(struct smp_shell_data *data, uint8_t byte)
 	}
 
+	if (IS_ENABLED(CONFIG_SHELL_LOG_BACKEND)) {
+		const struct shell *const sh = shell_backend_uart_get_ptr();
+
+		/* Reactivating the shell_log_backend */
+		z_shell_log_backend_resume(sh->log_backend);
+	}
```
What if the logging was already disabled? This code seems to enable it regardless if it was disabled by this code or not. The way we currently deal with this is we send "log halt" over the UART prior to using mcumgr. Also what sort of impact does this have performance wise vs without this code doing the log halt method, are you able to do a firmware update with both methods and measure the difference in time/speed?

sw commented:
Instead of making the logging system more complicated on the Zephyr side, maybe a better approach would be to make the mcumgr CLI tools more robust, so it can deal with non-SMP data?

de-nordic commented:
> 
> 
> Instead of making the logging or mcumgr subsystems more complicated on the Zephyr side, maybe a better approach would be to make the mcumgr CLI tools more robust, so it can deal with non-SMP data?

How would you tell the data apart? The whole upper/lowercase alphabet  consists of valid symbols for logging and for Base64 encoding, that is used in image transfers.

sw commented:
> How would you tell the data apart?

Anything starts with 06 09 or 04 14, ends with a newline and has a valid length and CRC should be considered an SMP fragment. Anything else can be discarded or handled as some other data such as logging data.

The Zephyr side (=SMP server) does that already, in `smp_shell_rx_bytes`. If you wanted a MCU implementation of the SMP client side, you could use that same logic. Though admittedly that function cannot handle other data before the start of an SMP fragment.

The mcumgr CLI suffers from the same issue, which is why I suggested an improvement there. Unfortunately my Go knowledge is nonexistent, so I can't really be of much help there.

The problem I see with disabling logging temporarily is a race condition where a log message is output just after the mcumgr-cli has written to the serial port and is now listening, but logging has not yet been disabled on the device.
